### ISSUE TITLE
[BUG] SyntaxTreeVisitor produces invalid node context values for Groups containing multiple Handlers

### ISSUE THREAD

**Issue by lukaskollmer** [Apodini/Apodini#111]
**Describe the bug**

For the following input web service:
```swift
struct WebService: Apodini.WebService {
    struct MyHandler: Handler {
        typealias Response = Never // make it crash if accessed
    }
    
    var content: some Component {
        Group("api") {
            Group("user") {
                MyHandler().operation(.read)
                MyHandler().operation(.create)
            }
        }
    }
}
```

the exported endpoints are:
```
/ {
  user/ {
    v1/ {
      api/ {
        - read: MyHandler() [0:0:0]
        - create: MyHandler() [0:0:1]
      }
    }
  }
}
```

This seems to be caused by putting multiple Handlers with different Operations into the same group.

It doesn't seem to be related to the nesting, this example also results in wrong paths:
```swift
var content: some Component {
    Group("api") {
        MyHandler().operation(.read)
        MyHandler().operation(.create)
    }
}
```

<br>

If, however, `WebService.content` is defined like this:
```swift
var content: some Component {
    Group("api") {
        Group("user") {
            MyHandler().operation(.read)
        }
    }
    Group("api") {
        Group("user") {
            MyHandler().operation(.create)
        }
    }
}
```
we're getting the expected output:
```
/ {
  v1/ {
    api/ {
      user/ {
        - create: MyHandler() [1:0:0]
        - read: MyHandler() [0:0:0]
      }
    }
  }
}
```

**To Reproduce**
Clone [`lukaskollmer/rest-path-bug`](https://github.com/Apodini/Apodini/tree/lukaskollmer/rest-path-bug) and run the TestWebService

**Comment by bauer-andreas** [comment]
Researched the issue a bit. I could isolate the issue that it is not in REST exporter nor in the Shared Model Builder.

I added the following print statement to the `SharedSemanticModelBuilder.register` method before the `Endpoint` is inserted into the tree.
```swift
print("Creating endpoint \(H.self) at path: \(paths)")
```

For your example at the top it prints:
```shell
Creating endpoint Text at path: ["user", Apodini.Version(prefix: "v", major: 1, minor: 0, patch: 0), "api"]
Creating endpoint Text at path: ["user", Apodini.Version(prefix: "v", major: 1, minor: 0, patch: 0), "api"]
```

Those paths are already messed up before they reach the `SharedSemanticModelBuilder`, thus this must happen within the scope of the `SyntaxTreeVisitor`.

**Comment by bauer-andreas** [comment]
Just printed out the `WebService.content` before create the `SyntaxTreeVisitor`. Just though of sharing it, as having the overview of the parsed structure might help(?)

```swift
Group<Group<TupleComponent<(OperationModifier<Text>, OperationModifier<Text>)>>>(
    pathComponents: ["api"],
    content: Apodini.Group<Apodini.TupleComponent<(Apodini.OperationModifier<Apodini.Text>, Apodini.OperationModifier<Apodini.Text>)>>(
        pathComponents: ["user"],
        content: Apodini.TupleComponent<(Apodini.OperationModifier<Apodini.Text>, Apodini.OperationModifier<Apodini.Text>)>(
            storage: (
                Apodini.OperationModifier<Apodini.Text>(
                    component: Apodini.Text(text: "user read"),
                    operation: Apodini.Operation.read
                ),
                Apodini.OperationModifier<Apodini.Text>(
                    component: Apodini.Text(text: "user create"),
                    operation: Apodini.Operation.create
                )
            ),
            file: "/Users/andi/XcodeProjects/TUM/Apodini/Sources/Apodini/Components/ComponentBuilder.swift",
            function: "buildBlock(_:_:)"
        )
    )
)
```

To my knowledge the `ComponentBuilder` operates as expected, thus the issue must be somewhere in parsing in the `SyntaxTreeVisitor`.

EDIT: I replaced the MyHandler with `Text("user read")` and `Text("user create")` respectively.

**Comment by lukaskollmer** [comment]
I tested this with a couple of old revisions, and determined that this bug was introduced in https://github.com/Apodini/Apodini/commit/da74f689294f09841b7e2ca8cf8dcbfba8d1af1a (and the associated PR #88).

The fix is relatively simple:
```diff
diff --git a/Sources/Apodini/Components/Group.swift b/Sources/Apodini/Components/Group.swift
index fae00be..998f8fb 100644
--- a/Sources/Apodini/Components/Group.swift
+++ b/Sources/Apodini/Components/Group.swift
@@ -24,19 +24,11 @@ public struct Group<Content: Component>: Component, SyntaxTreeVisitable {
     }
     
     func accept(_ visitor: SyntaxTreeVisitor) {
-        func forwardToContent() {
-            visitor.addContext(PathComponentContextKey.self, value: pathComponents, scope: .environment)
-            content.accept(visitor)
-        }
-        
-        if !String(describing: type(of: content)).hasPrefix("TupleComponent<") {
-            visitor.enterContent {
-                visitor.enterComponentContext {
-                    forwardToContent()
-                }
+        visitor.enterContent {
+            visitor.enterComponentContext {
+                visitor.addContext(PathComponentContextKey.self, value: pathComponents, scope: .environment)
+                content.accept(visitor)
             }
-        } else {
-            forwardToContent()
         }
     }
 }
```

@PSchmiedmayer I think this change was related to the "Reduce the usage of `enterComponentContext` and therefore reduce the depth of the syntax tree by excluding the double registration in `Group` and `TupleView`" part of the PR. Do you have any idea why this would cause the node context to get messed up?

The weird thing here is that [`TupleComponent.accept(_:)@da74f68`](https://github.com/Apodini/Apodini/blob/da74f689294f09841b7e2ca8cf8dcbfba8d1af1a/Sources/Apodini/Components/TupleComponent.swift#L24) also calls both `enterContent` (once for the entire tuple) and `enterComponentContext` (once per element in the tuple). My assumption would've been that, since the TupleComponent also calls these functions, removing them in the Group shouldn't have any impact?

**Comment by bauer-andreas** [comment]
To my understanding: The problem is having a `Group` (a) in a `Group` (b) in a `TupleComponent` (c).

In the visit of (b), it will only execute `forwardToContent()` - as the content is (c) - and thus not create a own `ContextNode` for (b). Meaning (b) will work on the `ContextNode` of (a). With the nature of [addContext](https://github.com/Apodini/Apodini/blob/e50376921981010f688d9c7e38ad4eb853f7cf2d/Sources/Apodini/SyntaxTreeVisitor/ContextNode.swift#L47-L61) this will **prepend** the PathComponents of (b) the path components of (a).

> It doesn't seem to be related to the nesting, this example also results in wrong paths:

This isn't entirely obvious with your second example WebService above, as the outer Group is hidden inside the WebService (what is creating the `v1` path).

**Comment by lukaskollmer** [comment]
Ooh ok so it's only indirectly related to having multiple Handlers in the group. (Since a group w/ a single child won't have its content wrapped in a Tuple, whereas one w/ multiple children will.)

**Comment by bauer-andreas** [comment]
Exactly, only if the content is a `TupleComponent`, the special case where `Group` operates on the `ContextNode` of the parent `Component` (to the Group) is executed. And the issue will only surface when the parent is actually a `Group` as then we have two Components working on the same `ContextKey` (`PathComponentContextKey`). 

I think we just should incorporate the fix you suggested (with additional regression testing). If I remember correctly, we already tried a similar improvement in a legacy version which still used `ComponentCollection`s and we failed (#13).
Every `Component` should have its own `ContextNode`, thus I think it makes sense that every Group gets its own as well(?)

**Comment by bauer-andreas** [comment]
@lukaskollmer can we turn your proposed fix into a PR?

**Comment by lukaskollmer** [comment]
Yeah I see no reason why not. I wanted to wait for Paul's opinion on this, but since he didn't comment I'll just assume he doesn't see any issues with this fix.

I'll open a PR and add him as reviewer.

**Comment by PSchmiedmayer** [comment]
@lukaskollmer @Supereg I just caught up with the discussion here while reading through the PRs and issues of the last few days. This is a very interesting issue, something I clearly didn't think about when writing the code.

I think the solution proposed by @lukaskollmer is a good fix the the problem, on the other hand it creates an additional nesting when talking about `HandlerIdentifiers` as `TupleComponent`s are no longer transparent to the algorithm. It would be interesting if we could achieve this transparency and also cover this edge case. If not doable in a reasonable fix that doesn't require us to rewrite the `SyntaxTreeVisitor` I would also be fine with merging the simple fix.

Feel free to create a PR and add me as a reviewer! 🚀

**Comment by lukaskollmer** [comment]
@PSchmiedmayer I'll have a look at maybe adjusting the endpoint identifier algorithm to support this, but tbh I don't think it really matters that much for now, since the actual values of these identifiers are never exposed to the user.

We currently don't really care about an endpoint identifier's raw value, the only thing we care about is that they are stable across multiple executions of the web service.

**Comment by PSchmiedmayer** [comment]
@lukaskollmer I agree with your argumentation, you are right. As these identifiers are just internal this would be fine 👍

**Comment by lschlesinger** [comment]
Might be still a problem, as the `endpoint.identifier.rawValue` on 548787e still has some invalid (negative) values. It seems that the third number is always `-1`, e.g for `TraditionalGreeter` it is `0:2:-1:0`.

**Comment by lukaskollmer** [comment]
fixed again, this time via #187
